File: day-09/part-2/silvestre.cs
using System;
using System.Collections.Generic;


namespace Aoc
{
    class Solution
    {
        private const ushort LINE_LENGTH = 100+1;
        private const ushort N_LINES = 100;
        private static int BassinSize(char[] input, ushort cursor, ushort col) {
            int size = 0;
            ushort current;
            Stack<ushort> toVisit = new Stack<ushort>();
            byte[] visited = new byte[N_LINES*LINE_LENGTH];
            toVisit.Push(cursor);
            while (toVisit.Count > 0) {
                if (toVisit.Count == 0) continue;
                current = toVisit.Pop();
                if (visited[current] == 1 || input[current] == '\n' || input[current] == '9') {
                    visited[current] = 1;
                    continue;
                } else {
                    ++size;
                    visited[current] = 1;
                    if (current >= LINE_LENGTH && visited[current-LINE_LENGTH] == 0) {
                        toVisit.Push((ushort)(current-LINE_LENGTH));
                    }
                    if (current < LINE_LENGTH * (N_LINES-1) && visited[current+LINE_LENGTH] == 0) {
                        toVisit.Push((ushort)(current+LINE_LENGTH));
                    }
                    if ((current < LINE_LENGTH * N_LINES - 2) && visited[current+1] == 0) {
                        toVisit.Push((ushort)(current+1));
                    }
                    if (current > 0  && visited[current-1] == 0) {
                        toVisit.Push((ushort)(current-1));
                    }
                }
            }
            return size;
        }
        private static bool IsLowPoint(char[] input, ushort cursor, ushort col) {
            return (
                ((cursor < LINE_LENGTH) || (input[cursor] < input[cursor-LINE_LENGTH])) &&
                ((cursor >= LINE_LENGTH * (N_LINES-1)) || (input[cursor] < input[cursor+LINE_LENGTH])) &&
                ((col == 1) || (input[cursor] < input[cursor-1])) &&
                ((col == LINE_LENGTH-1) || (input[cursor] < input[cursor+1]))
            );
        }
        private static int Solve(char[] input) {
            int top1, top2, top3, current;
            top1 = top2 = top3 = current = 0;
            ushort cursor = 0;
            ushort col = 1;
            while (cursor < LINE_LENGTH * N_LINES - 1) {
                if (input[cursor] == '\n') {col = 0;}
                else if (IsLowPoint(input, cursor, col)) {
                    current = BassinSize(input, cursor, col);
                    if (current > top1) {top3 = top2; top2 = top1; top1 = current;}
                    else if (current > top2) {top3 = top2; top2 = current;}
                    else if (current > top3) {top3 = current;}
                }
                ++col;
                ++cursor;
            }
            return top1 * top2 * top3;
        }

        public static void Main(string[] args) {
            char[] input = args[0].ToCharArray();
            var watch = new System.Diagnostics.Stopwatch();
            watch.Start();
            int result = Solve(input);
            watch.Stop();
            Console.WriteLine("_duration: " + watch.Elapsed.TotalMilliseconds + "\n" + result);
        }
        
    }
}

File: day-09/part-2/silvestre.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <stdbool.h>

struct stack {
    unsigned short array[128];
    unsigned short length;
};

void push(struct stack* s, unsigned short value) {
    s->length++; s->array[s->length-1] = value; 
}

unsigned short pop(struct stack* s) {
    s->length--; 
    return s->array[s->length]; 
}

const short NUM_LINES = 100;
const short LINE_LENGTH = 100+1;

bool is_low_point(char* s, unsigned short* p_cursor, unsigned short* p_col) {
    return (*p_cursor < LINE_LENGTH || s[*p_cursor] < s[*p_cursor - LINE_LENGTH]) &&
        (*p_cursor >  LINE_LENGTH * (NUM_LINES - 1) || s[*p_cursor] < s[*p_cursor + LINE_LENGTH]) &&
        (*p_col == 1 || s[*p_cursor] < s[*p_cursor - 1]) &&
        (*p_col == LINE_LENGTH - 1 || s[*p_cursor] < s[*p_cursor + 1]);
}

unsigned long bassin_size(char* s, char* visited, struct stack* to_visit, unsigned short cursor) {
    unsigned long size = 0;
    unsigned short current;
    push(to_visit, cursor);
    while (to_visit->length) {
        current = pop(to_visit);
        if (!visited[current] &&  s[current] != '\n' && s[current] != '9') {
            size++;
            if (current >= LINE_LENGTH && !visited[current - LINE_LENGTH]) {
                push(to_visit, current - LINE_LENGTH);
            }
            if (current < LINE_LENGTH * (NUM_LINES - 1) && !visited[current + LINE_LENGTH]) {
                push(to_visit, current + LINE_LENGTH);
            }
            if (current < LINE_LENGTH * NUM_LINES - 2 && !visited[current + 1]) {
                push(to_visit, current + 1);
            }
            if (current > 0 && !visited[current - 1]) {
                push(to_visit, current - 1);
            }
        }
        visited[current] = 1;
    }
    return size;
}

unsigned long run(char* s) {
    // Your code goes here
    unsigned long top1, top2, top3, current;
    top1 = top2 = top3 = current = 0;
    char visited[100*101] = {0};
    struct stack to_visit;
    to_visit.length = 0;
    unsigned short cursor = 0;
    unsigned short col = 1;
    while (cursor < NUM_LINES * LINE_LENGTH - 1) {
        if (s[cursor] == '\n') {
            col = 0;
        } else if (is_low_point(s, &cursor, &col)) {
            current = bassin_size(s, visited, &to_visit, cursor);
            if (current > top1) {top3 = top2; top2 = top1; top1 = current;}
            else if (current > top2) {top3 = top2; top2 = current;}
            else if (current > top3) {top3 = current;}
        }
        cursor++;
        col++;

    }
    return top1 * top2 * top3;
}

int main(int argc, char** argv)
{
    if (argc < 2) {
        printf("Missing one argument\n");
        exit(1);
    }

    clock_t start = clock();
    unsigned long answer = run(argv[1]);
    
    printf("_duration:%f\n%lu\n", (float)( clock () - start ) * 1000.0 /  CLOCKS_PER_SEC, answer);
    return 0;
}
